src/RADOS/RADOS.cc

```diff
@@ -0,0 +1,1356 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+
+#define BOOST_BIND_NO_PLACEHOLDERS
+
+#include <optional>
+#include <string_view>
+
+#include <boost/intrusive_ptr.hpp>
+
+#include <fmt/format.h>
+
+#include "include/ceph_fs.h"
+#include "common/ceph_context.h"
+#include "common/ceph_argparse.h"
+#include "common/common_init.h"
+#include "common/hobject.h"
+
+#include "global/global_init.h"
+
+#include "osd/osd_types.h"
+
+#include "RADOS/RADOSImpl.h"
+#include "include/RADOS/RADOS.hpp"
+
+namespace RADOS {
+// Object
+
+Object::Object(std::string_view s) {
+  static_assert(impl_size >= sizeof(object_t));
+  new (&impl) object_t(s);
+}
+
+Object::Object(std::string&& s) {
+  static_assert(impl_size >= sizeof(object_t));
+  new (&impl) object_t(std::move(s));
+}
+
+Object::Object(const std::string& s) {
+  static_assert(impl_size >= sizeof(object_t));
+  new (&impl) object_t(s);
+}
+
+Object::~Object() {
+  reinterpret_cast<object_t*>(&impl)->~object_t();
+}
+
+Object::Object(const Object& o) {
+  static_assert(impl_size >= sizeof(object_t));
+  new (&impl) object_t(*reinterpret_cast<const object_t*>(&o.impl));
+}
+Object& Object::operator =(const Object& o) {
+  *reinterpret_cast<object_t*>(&impl) =
+    *reinterpret_cast<const object_t*>(&o.impl);
+  return *this;
+}
+Object::Object(Object&& o) {
+  static_assert(impl_size >= sizeof(object_t));
+  new (&impl) object_t(std::move(*reinterpret_cast<object_t*>(&o.impl)));
+}
+Object& Object::operator =(Object&& o) {
+  *reinterpret_cast<object_t*>(&impl) =
+    std::move(*reinterpret_cast<object_t*>(&o.impl));
+  return *this;
+}
+
+Object::operator std::string_view() const {
+  return std::string_view(reinterpret_cast<const object_t*>(&impl)->name);
+}
+
+bool operator <(const Object& lhs, const Object& rhs) {
+  return (*reinterpret_cast<const object_t*>(&lhs.impl) <
+	  *reinterpret_cast<const object_t*>(&rhs.impl));
+}
+bool operator <=(const Object& lhs, const Object& rhs) {
+  return (*reinterpret_cast<const object_t*>(&lhs.impl) <=
+	  *reinterpret_cast<const object_t*>(&rhs.impl));
+}
+bool operator >=(const Object& lhs, const Object& rhs) {
+  return (*reinterpret_cast<const object_t*>(&lhs.impl) >=
+	  *reinterpret_cast<const object_t*>(&rhs.impl));
+}
+bool operator >(const Object& lhs, const Object& rhs) {
+  return (*reinterpret_cast<const object_t*>(&lhs.impl) >
+	  *reinterpret_cast<const object_t*>(&rhs.impl));
+}
+
+bool operator ==(const Object& lhs, const Object& rhs) {
+  return (*reinterpret_cast<const object_t*>(&lhs.impl) ==
+	  *reinterpret_cast<const object_t*>(&rhs.impl));
+}
+bool operator !=(const Object& lhs, const Object& rhs) {
+  return (*reinterpret_cast<const object_t*>(&lhs.impl) !=
+	  *reinterpret_cast<const object_t*>(&rhs.impl));
+}
+
+// IOContext
+
+struct IOContextImpl {
+  object_locator_t oloc;
+  snapid_t snap_seq;
```

Comment:
for some reason, writes are being issued with CEPH_NOSNAP and reads are defaulting to 0. initializing this to CEPH_NOSNAP seemed to fix

Reply:
cherry-picked